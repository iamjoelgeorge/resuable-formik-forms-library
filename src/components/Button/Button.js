import React from 'react';

import styles from './Button.module.scss';

const Button = (props) => {
  const { label, type, onClick, ...rest } = props;

  return (
    <button className={styles.button} type={type} {...rest}>
      Submit
    </button>
  );
};

export default Button;
