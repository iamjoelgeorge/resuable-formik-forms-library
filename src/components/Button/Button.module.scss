@import '../../theme/variables.scss';

.button {
  border: none;
  padding: 10px;
  border-radius: 2px;
  background-color: $darkPurple;
  color: white;
  font-weight: bold;
  margin-top: 12px;

  &:disabled {
    background-color: rgb(89, 89, 89);
  }
}
