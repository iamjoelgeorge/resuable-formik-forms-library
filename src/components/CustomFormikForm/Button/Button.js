import React from 'react';

import PropTypes from 'prop-types';
import { useFormikContext } from 'formik';

import styles from './Button.module.scss';
import { joinClassNames } from '../../../utils/utils';
import { buttonTheme, buttonTypes, buttonVariants } from '../../../constants/constants';
import {
  ExternalLinkIconWhite,
  ExternalLinkIconRed,
  ExternalLinkIconPurple,
} from '../../../constants/icons';

const Button = (props) => {
  const {
    variant,
    type,
    href,
    label,
    onClick,
    isDisabled,
    theme,
    showExternalLinkIcon,
    containerClass,
    ...rest
  } = props;

  const { isValid } = useFormikContext();

  // const disableButton = type === buttonTypes.submit ? !(formik.dirty && formik.isValid) : isDisabled;
  const disableButton = type === buttonTypes.submit ? !isValid : isDisabled;
  const goTo = !isDisabled ? href : '';

  const linkDisabledClass = isDisabled ? styles.linkDisabled : '';
  const linkAsButtonDisabledClass = isDisabled ? styles.linkButtonDisabled : '';

  const IconForLink = theme === buttonTheme.purple ? ExternalLinkIconPurple : ExternalLinkIconRed;

  const buttonClasses =
    theme === buttonTheme.purple
      ? joinClassNames([styles.button, styles.purpleTheme, containerClass])
      : joinClassNames([styles.button, styles.redTheme, containerClass]);

  const linkClasses =
    theme === buttonTheme.purple
      ? joinClassNames([styles.link, styles.purpleTheme, linkDisabledClass, containerClass])
      : joinClassNames([styles.link, styles.redTheme, linkDisabledClass, containerClass]);

  const linkAsButtonClasses = joinClassNames([
    buttonClasses,
    styles.linkAsButton,
    linkAsButtonDisabledClass,
  ]);

  const handleClickOnLink = (e) => {
    if (isDisabled) e.preventDefault();
  };

  const renderButton = () => {
    switch (variant) {
      case buttonVariants.button:
        return (
          <button
            className={buttonClasses}
            type={type}
            onClick={onClick}
            disabled={disableButton}
            data-testid={`button-${label}`}
            {...rest}
          >
            {label}
          </button>
        );

      case buttonVariants.link:
        return (
          <a
            className={linkClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
            data-testid={`link-${label}`}
          >
            <span>{label}</span>
            {showExternalLinkIcon && (
              <img className={styles.externalLinkIcon} src={IconForLink} alt='External Link' />
            )}
          </a>
        );

      case buttonVariants.linkAsButton:
        return (
          <a
            className={linkAsButtonClasses}
            href={goTo}
            target='_blank'
            onClick={handleClickOnLink}
            rel='noreferrer'
            data-testid={`link_as_button-${label}`}
          >
            {label}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </a>
        );

      default:
        return (
          <button
            className={buttonClasses}
            type={type}
            disabled={disableButton}
            onClick={onClick}
            data-testid={`button-${label}`}
            {...rest}
          >
            {label}{' '}
            {showExternalLinkIcon && (
              <img
                className={styles.externalLinkIcon}
                src={ExternalLinkIconWhite}
                alt='External Link'
              />
            )}
          </button>
        );
    }
  };

  return <>{renderButton()}</>;
};

Button.propTypes = {
  variant: PropTypes.oneOf([
    buttonVariants.button,
    buttonVariants.link,
    buttonVariants.linkAsButton,
  ]),
  type: PropTypes.oneOf([buttonTypes.submit, buttonTypes.button]),
  href: PropTypes.string,
  label: PropTypes.string,
  onClick: PropTypes.func,
  isDisabled: PropTypes.bool,
  theme: PropTypes.oneOf([buttonTheme.red, buttonTheme.purple]),
  showExternalLinkIcon: PropTypes.bool,
  containerClass: PropTypes.string,
};

Button.defaultProps = {
  variant: buttonVariants.button,
  type: buttonTypes.submit,
  href: '#',
  label: '',
  onClick: () => {},
  isDisabled: false,
  theme: buttonTheme.purple,
  showExternalLinkIcon: false,
  containerClass: '',
};

export default Button;
