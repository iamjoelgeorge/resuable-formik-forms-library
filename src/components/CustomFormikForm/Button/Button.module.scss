@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

* {
  box-sizing: border-box;
}

@mixin useButtonTheme($bgColor, $hoverBgColor, $focusBgColor) {
  background-color: $bgColor;

  &:hover:not(:disabled),
  &:active {
    cursor: pointer;
    background-color: $hoverBgColor;
  }

  &:focus {
    border-color: $focusBgColor;
  }
}

@mixin useLinkTheme(
  $defaultTextColor,
  $hoverColor,
  $borderBottomColorOnFocus,
  $borderBottomColorOnActive,
  $textColorOnActive
) {
  color: $defaultTextColor;

  span {
    border: 2px solid transparent;
    padding-bottom: 2px;
  }

  &:hover:not(&.linkDisabled) {
    color: $hoverColor;
  }

  &:focus {
    span {
      border-bottom-style: dashed;
      border-bottom-color: $borderBottomColorOnFocus;
    }
  }

  &:active {
    border-color: $borderBottomColorOnActive;
    color: $textColorOnActive;
  }
}

.button,
.link {
  border: 3px solid transparent;
  border-radius: 4px;
  font-size: em(16);
  font-weight: $fontWeightExtraBold;
  line-height: 19px;
  margin-top: 12px;
  padding: 0 em(15);
  width: 198px;
  height: 51px;
  text-align: center;
}

.button {
  color: $white;
  transition: 0.1s all ease-in-out;
  outline: none;

  &.purpleTheme {
    @include useButtonTheme($primary, $darkPurple, $brightPurple);
  }

  &.redTheme {
    @include useButtonTheme($tooltipRed, $buttonHover, $pinkOne);
  }

  &:disabled {
    background-color: $white;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.linkAsButton {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;

  &.linkButtonDisabled {
    background-color: $white !important;
    cursor: not-allowed !important;
    color: $tertiaryGrey;
    border-color: $dGrey;
  }
}

.link {
  display: flex;
  justify-content: center;
  align-items: center;
  text-decoration: none;
  outline: none;

  &.purpleTheme {
    @include useLinkTheme($primary, $darkPurple, $darkPurple, $brightPurple, $darkPurple);
  }

  &.redTheme {
    @include useLinkTheme($tooltipRed, $buttonHover, $tooltipRed, $pinkOne, $buttonHover);
  }

  &.linkDisabled {
    color: $tertiaryGrey;
    pointer-events: none;

    .externalLinkIcon {
      -webkit-filter: invert(50%); /* safari 6.0 - 9.0 */
      filter: invert(50%);
    }
  }
}

.externalLinkIcon {
  margin-left: em(8);
}
