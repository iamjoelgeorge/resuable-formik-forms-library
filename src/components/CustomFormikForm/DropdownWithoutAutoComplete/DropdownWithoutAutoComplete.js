import React, { useRef } from 'react';

import { Field } from 'formik';
import { v4 as uuidv4 } from 'uuid';
import PropTypes from 'prop-types';

import styles from './DropdownWithoutAutoComplete.module.scss';
import { ArrowNext } from '../../../constants/icons';
import { joinClassNames } from '../../../utils/utils';
import { useToggleDropdown } from '../../../hooks/useToggleDropdown';
import SlidingLabel from '../SlidingLabel/SlidingLabel';
import ErrorText from '../ErrorText/ErrorText';
import AdditionalInfo from '../AdditionalInfo/AdditionalInfo';

const DropdownWithoutAutoComplete = (props) => {
  const {
    name,
    label,
    dropdownArray,
    containerClass: customContainerClass,
    labelTooltipBoxHeading,
    labelTooltipBoxDescription,
    labelTooltipBoxDescriptionElement,
    tooltipLink,
    tooltipLinkText,
    helpLinkText,
    helpLink,
    optionalText,
    isRequired,
    isDisabled,
    formik,
    ...rest
  } = props;

  const { errors, values } = formik;

  const dropdownContainerRef = useRef();
  const [isDropdownOpen, setIsDropdownOpen, toggleDropdown] =
    useToggleDropdown(dropdownContainerRef);

  const userHasVisitedTheInputField = formik.touched[name];
  const inputFieldHasErrors = errors[name];
  const addErrorClassesToLabelAndInput = !!userHasVisitedTheInputField && !!inputFieldHasErrors;

  const containerClasses = isDisabled
    ? joinClassNames([styles.container, styles.isDisabled, customContainerClass])
    : joinClassNames([styles.container, customContainerClass]);

  const renderDropdownItems = (setFieldValue) =>
    dropdownArray?.map((item) => {
      const dropdownItemClasses =
        String(item).trim().toLowerCase() === values[name].trim().toLowerCase()
          ? joinClassNames([styles.dropdownItem, styles.selected])
          : styles.dropdownItem;

      return (
        <li key={uuidv4()} className={dropdownItemClasses}>
          <button
            data-testid={`${name}-dropdown-item`}
            type='button'
            onClick={() => setFieldValue(name, item)}
          >
            {item}
          </button>
        </li>
      );
    });

  return (
    <Field name={name} {...rest}>
      {({ form, field }) => {
        const { setFieldValue } = form;

        return (
          <div data-testid={`${name}-dropdown-component`} className={containerClasses}>
            {label && (
              <SlidingLabel
                label={label}
                inputEntered={!!field.value}
                htmlFor={'selectedDate'}
                showErrorStyle={addErrorClassesToLabelAndInput}
                tooltipBoxHeading={labelTooltipBoxHeading}
                tooltipBoxDescription={labelTooltipBoxDescription}
                tooltipBoxDescriptionElement={labelTooltipBoxDescriptionElement}
                inputIsRequired={isRequired}
              />
            )}
            <div
              role='button'
              data-testid={`${name}-dropdown-button`}
              tabIndex={!isDisabled ? '0' : ''}
              ref={dropdownContainerRef}
              className={styles.dropdownContainer}
              onClick={!isDisabled ? toggleDropdown : () => {}}
            >
              {field?.value}

              <span className={styles.dropdownIconContainer}>
                <img src={ArrowNext} alt='Dropdown icon' />
              </span>

              {isDropdownOpen && (
                <ul data-testid={`${name}-dropdown-list`} className={styles.dropdownList}>
                  {renderDropdownItems(setFieldValue)}
                </ul>
              )}
            </div>

            {errors[name] && formik.touched[name] && (
              <ErrorText containerClass={styles.errorContainer} fieldName={name} />
            )}
            <AdditionalInfo
              optionalText={optionalText}
              helpLinkText={helpLinkText}
              helpLink={helpLink}
              tooltipLinkText={tooltipLinkText}
              tooltipLink={tooltipLink}
            />
          </div>
        );
      }}
    </Field>
  );
};

DropdownWithoutAutoComplete.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string,
  dropdownArray: PropTypes.array.isRequired,
  formik: PropTypes.shape({}),
  containerClass: PropTypes.string,
  labelTooltipBoxHeading: PropTypes.string,
  labelTooltipBoxDescription: PropTypes.string,
  labelTooltipBoxDescriptionElement: PropTypes.element,
  tooltipLink: PropTypes.string,
  tooltipLinkText: PropTypes.string,
  helpLinkText: PropTypes.string,
  helpLink: PropTypes.string,
  optionalText: PropTypes.string,
  isDisabled: PropTypes.bool,
  isRequired: PropTypes.bool,
};

DropdownWithoutAutoComplete.defaultProps = {
  label: '',
  formik: {},
  containerClass: '',
  labelTooltipBoxHeading: '',
  labelTooltipBoxDescription: '',
  labelTooltipBoxDescriptionElement: null,
  tooltipLink: '',
  tooltipLinkText: '',
  helpLinkText: '',
  helpLink: '',
  optionalText: '',
  isDisabled: false,
  isRequired: false,
};

export default DropdownWithoutAutoComplete;
