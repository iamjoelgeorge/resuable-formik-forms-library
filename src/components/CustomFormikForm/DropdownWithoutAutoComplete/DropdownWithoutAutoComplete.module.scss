@import '../../../theme/variables.scss';
@import '../../../theme/utils.scss';

* {
  box-sizing: border-box;
}

.container {
  position: relative;

  .dropdownContainer {
    position: relative;
    display: flex;
    align-items: center;
    width: 276px;
    max-width: 100%;
    min-height: em(24);
    background-color: transparent;
    color: $primaryDark;
    border: 0;
    border-bottom: 1px solid $lighterGrey;
    padding: 0 em(20) em(6) em(10);
    font-size: 14px;
    margin: 0;
    transition: 0.1s all ease-in-out;
    outline: none;

    &:hover:not(:disabled),
    &:focus {
      cursor: pointer;
      border-bottom-color: $brightPurpleTwo;
    }

    &:focus {
      border-bottom-style: dashed;
    }

    .dropdownIconContainer {
      position: absolute;
      right: -3px;
      width: 10px;
      height: 10px;
      transform: rotate(90deg);

      img {
        width: 100%;
        height: 100%;
        object-fit: contain;
      }
    }

    ul.dropdownList {
      list-style: none;
      position: absolute;
      left: 0;
      top: 95%;
      width: 100%;
      margin: 0;
      padding: 0;
      box-shadow: 0 2px 5px 2px $borderGrey;
      z-index: 1;
      background-color: $white;
      border-top: 1px solid $primary;
      max-height: 125px;
      overflow-y: scroll;

      /* width */
      &::-webkit-scrollbar {
        width: 2px;
      }

      /* Handle */
      &::-webkit-scrollbar-thumb {
        background: $lighterDarkGrey;
        border-radius: 10px;
      }

      li.dropdownItem {
        button {
          background-color: $white;
          padding: 5px 10px;
          width: 100%;
          cursor: pointer;
          border: 1px solid transparent;
          transition: 0.1s all ease-in-out;
          font-size: 14px;
          outline: none;
          text-align: left;

          &:hover,
          &:focus {
            background-color: $lightPurple;
          }
        }
        &.selected {
          button {
            color: $primary;
            font-weight: $fontWeightBold;
          }
        }
      }
    }
  }

  &.isDisabled {
    label {
      color: $secondaryGrey;
    }

    .dropdownContainer {
      color: $tertiaryGrey;

      &:hover {
        border-bottom-color: $lighterGrey;
        cursor: not-allowed;
      }
    }

    .dropdownIconContainer {
      -webkit-filter: invert(50%); /* safari 6.0 - 9.0 */
      filter: invert(50%);
      opacity: 0.3;
    }
  }

  .errorContainer {
    padding-left: em(14);
  }
}
