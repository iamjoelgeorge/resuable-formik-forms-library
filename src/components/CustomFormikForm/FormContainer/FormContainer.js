import React, { Children, cloneElement } from 'react';

import PropTypes from 'prop-types';
import { Formik, Form } from 'formik';
import * as Yup from 'yup';

import './FormContainer.module.scss';
import { validateInput } from '../../../utils/validateInput';

const FormContainer = (props) => {
  const { initialValues, children, validations, containerClass, onSubmit, ...rest } = props;

  const validationObject = {};
  validations.forEach((item) => {
    validateInput(item, validationObject);
  });

  const validationSchema = Yup.object().shape(validationObject);

  const renderChildren = (child, formik) => {
    const childOfChild = child.props.children;
    const { className } = child.props;

    if (childOfChild?.props?.children) {
      return renderChildren(childOfChild, formik);
    }

    return (
      <div className={className}>
        {childOfChild?.length
          ? /*
            The error generated because of the key not being added can be ignored in this case
            because these elements will not be deleted/modified. The key is not added the the child
            because it is interfering with the focus event of the input field.
          */
            childOfChild?.map((childElement) => cloneElement(childElement, { formik }))
          : cloneElement(childOfChild, { formik })}
      </div>
    );
  };
  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema}>
      {(formik) => {
        const childrenWithExtraProp = Children.map(children, (child) => {
          return child?.props?.children
            ? renderChildren(child, formik)
            : cloneElement(child, { formik });
        });
        return (
          <Form className={containerClass} {...rest}>
            {childrenWithExtraProp}
          </Form>
        );
      }}
    </Formik>
  );
};

FormContainer.propTypes = {
  initialValues: PropTypes.shape({}),
  children: PropTypes.element,
  validations: PropTypes.array,
  containerClass: PropTypes.string,
  onSubmit: PropTypes.func,
};

FormContainer.defaultProps = {
  initialValues: {},
  children: null,
  validations: [],
  containerClass: '',
  onSubmit: () => {},
};

export default FormContainer;
