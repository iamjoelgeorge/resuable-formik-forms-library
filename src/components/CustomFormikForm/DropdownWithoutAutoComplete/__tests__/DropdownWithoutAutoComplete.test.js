/* eslint-disable */
import React from 'react';

import { render, cleanup, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

import Button from '../../Button/Button';
import FormContainer from '../../FormContainer/FormContainer';
import DropdownWithoutAutoComplete from '../DropdownWithoutAutoComplete';

const defaultProps = {
  name: 'salutation',
  label: 'salutation',
  dropdownArray: ['Other', 'Mr.', 'Ms.', 'Mrs.'],
  containerClass: '',
  labelTooltipBoxHeading: '',
  labelTooltipBoxDescription: '',
  labelTooltipBoxDescriptionElement: null,
  tooltipLink: '',
  tooltipLinkText: '',
  helpLinkText: '',
  helpLink: '',
  optionalText: '',
  isDisabled: false,
  isRequired: false,
};

const initialValues = {
  salutation: '',
};

const defaultValidations = [];
const handleSubmit = jest.fn();

afterEach(cleanup);

describe('[Component]: DropdownWithoutAutoComplete', () => {
  it('should match the Snapshot', () => {
    const tree = renderer
      .create(
        <FormContainer
          initialValues={initialValues}
          validations={defaultValidations}
          onSubmit={handleSubmit}
        >
          <DropdownWithoutAutoComplete {...defaultProps} />
        </FormContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render the component correctly', () => {
    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...defaultProps} />
      </FormContainer>,
    );

    expect(getByTestId('salutation-dropdown-component')).toBeTruthy();
  });

  it('should render a disabled dropdown component', () => {
    const props = {
      ...defaultProps,
      isDisabled: true,
    };

    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    expect(getByTestId('salutation-dropdown-button')).toHaveAttribute('tabIndex', '');
  });

  it('should render the optional text', () => {
    const props = {
      ...defaultProps,
      optionalText: 'Test Optional text',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    expect(getByText('Test Optional text')).toBeTruthy();
  });

  it('should render the help link', () => {
    const props = {
      ...defaultProps,
      helpLinkText: 'Test help link',
      helpLink: 'https://virginaustralia.com/',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    expect(getByText('Test help link')).toBeTruthy();
  });

  it('should render the tooltip link', () => {
    const props = {
      ...defaultProps,
      tooltipLinkText: 'Test tooltip link',
      tooltipLink: 'https://virginaustralia.com/',
    };

    const { getByText } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    expect(getByText('Test tooltip link')).toBeTruthy();
  });

  it('should render a default value', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const initialValues = {
      salutation: 'Mr.',
    };

    const validations = [
      {
        name: 'salutation',
        type: 'dropdown',
        isRequired: true,
        message: 'Please select how you would like to be addressed.',
      },
    ];

    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    expect(getByTestId('salutation-dropdown-button')).toHaveTextContent('Mr.');
  });

  it('should render the dropdown list', async () => {
    const props = {
      ...defaultProps,
    };

    const { getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    const dropdownButton = getByTestId('salutation-dropdown-button');
    userEvent.click(dropdownButton);

    await waitFor(() => {
      expect(getByTestId('salutation-dropdown-list')).toBeTruthy();
    });
  });

  it('should render the selected option', async () => {
    const props = {
      ...defaultProps,
    };

    const { getByTestId, getAllByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={defaultValidations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
      </FormContainer>,
    );

    const dropdownButton = getByTestId('salutation-dropdown-button');
    userEvent.click(dropdownButton);

    await waitFor(() => {
      const dropdownList = getByTestId('salutation-dropdown-list');
      expect(dropdownList).toBeTruthy();
    });

    let options = getAllByTestId('salutation-dropdown-item');
    userEvent.click(options[0]);

    await waitFor(() => {
      expect(dropdownButton).toHaveTextContent('Other');
    });

    userEvent.click(dropdownButton);

    await waitFor(() => {
      const dropdownList = getByTestId('salutation-dropdown-list');
      expect(dropdownList).toBeTruthy();
    });

    options = getAllByTestId('salutation-dropdown-item');
    userEvent.click(options[2]);

    await waitFor(() => {
      expect(dropdownButton).toHaveTextContent('Ms.');
    });
  });

  it('should show the error component for a required Dropdown', async () => {
    const props = {
      ...defaultProps,
      isRequired: true,
    };

    const validations = [
      {
        name: 'salutation',
        type: 'dropdown',
        isRequired: true,
        message: 'Please select how you would like to be addressed.',
      },
    ];

    const { getByText, getByTestId } = render(
      <FormContainer
        initialValues={initialValues}
        validations={validations}
        onSubmit={handleSubmit}
      >
        <DropdownWithoutAutoComplete {...props} />
        <Button label='Submit' />
      </FormContainer>,
    );

    const submitButton = getByText('Submit');
    userEvent.click(submitButton);

    await waitFor(() => {
      expect(getByTestId('error-salutation')).toHaveTextContent(
        'Please select how you would like to be addressed.',
      );
    });
  });
});
