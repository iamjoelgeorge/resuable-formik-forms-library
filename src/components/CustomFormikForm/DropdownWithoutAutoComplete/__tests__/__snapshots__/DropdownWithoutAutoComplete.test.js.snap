// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component]: DropdownWithoutAutoComplete should match the Snapshot 1`] = `
<form
  action="#"
  className=""
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="container "
    data-testid="salutation-dropdown-component"
  >
    <div
      className="container  "
    >
      <label
        className="label"
        htmlFor="selectedDate"
      >
        salutation
      </label>
    </div>
    <div
      className="dropdownContainer"
      data-testid="salutation-dropdown-button"
      onClick={[Function]}
      role="button"
      tabIndex="0"
    >
      
      <span
        className="dropdownIconContainer"
      >
        <img
          alt="Dropdown icon"
          src="test-file-stub"
        />
      </span>
    </div>
    
    
    
  </div>
</form>
`;
